scylla-cdc/src/consumer.rs

```diff
@@ -0,0 +1,474 @@
+        //          to something other than a hardcoded tuple.
+
+        let mut stream_id = 0;
+        let mut time = 0;
+        let mut batch_seq_no = 0;
+        let mut end_of_batch = 0;
+        let mut operation = 0;
+        let mut ttl = 0;
+        let mut mapping: HashMap<String, usize> = HashMap::new();
+        let mut deleted_mapping: HashMap<String, usize> = HashMap::new();
+        let mut deleted_el_mapping: HashMap<String, usize> = HashMap::new();
+
+        let mut j = 0;
+
+        // Hashmaps will have indices of data in a new vector without the hardcoded values.
+        for (i, spec) in specs.iter().enumerate() {
+            match spec.name.as_str() {
+                // spec.name is public since 0.3.0 driver version, it didn't work on 0.2.1.
+                STREAM_ID_NAME => stream_id = i,
+                TIME_NAME => time = i,
+                BATCH_SEQ_NO_NAME => batch_seq_no = i,
+                END_OF_BATCH_NAME => end_of_batch = i,
+                OPERATION_NAME => operation = i,
+                TTL_NAME => ttl = i,
+                x => {
+                    if let Some(stripped) = x.strip_prefix(ARE_ELEMENTS_DELETED_PREFIX) {
+                        deleted_el_mapping.insert(stripped.to_string(), j);
+                    } else if let Some(stripped) = x.strip_prefix(IS_DELETED_PREFIX) {
+                        deleted_mapping.insert(stripped.to_string(), j);
+                    } else {
+                        mapping.insert(x.to_string(), j);
+                    }
+
+                    j += 1;
+                }
+            }
+        }
+
+        CDCRowSchema {
+            stream_id,
+            time,
+            batch_seq_no,
+            end_of_batch,
+            operation,
+            ttl,
+            mapping,
+            deleted_mapping,
+            deleted_el_mapping,
+        }
+    }
+}
+
+pub struct CDCRow<'schema> {
+    pub stream_id: Vec<u8>,
+    pub time: uuid::Uuid,
+    pub batch_seq_no: i32,
+    pub end_of_batch: bool,
+    pub operation: OperationType,
+    // Can be NULL in the database.
+    pub ttl: Option<i64>,
+    data: Vec<Option<CqlValue>>,
+    // Maps element name to its index in the data vector.
+    schema: &'schema CDCRowSchema,
+}
+
+impl CDCRow<'_> {
+    pub fn from_row(row: Row, schema: &CDCRowSchema) -> CDCRow {
+        // If cdc read was successful, these default values will not be used.
+        let mut stream_id = vec![];
+        let mut time = uuid::Uuid::default();
+        let mut batch_seq_no = i32::MAX;
+        let mut end_of_batch = false;
+        let mut operation = OperationType::PreImage;
+        let mut ttl = None;
+
+        let data_count =
+            schema.mapping.len() + schema.deleted_mapping.len() + schema.deleted_el_mapping.len();
+        let mut data: Vec<Option<CqlValue>> = Vec::with_capacity(data_count);
+
+        for (i, column) in row.columns.into_iter().enumerate() {
+            if i == schema.stream_id {
```

Comment:
Perhaps using guards would be more elegant:
https://doc.rust-lang.org/rust-by-example/flow_control/match/guard.html
(just a suggestion, depends on your preference)